source/extensions/filters/network/mysql_proxy/mysql_filter.cc

```diff
@@ -23,23 +23,26 @@ void MySQLFilter::initializeReadFilterCallbacks(Network::ReadFilterCallbacks& ca
 }
 
 Network::FilterStatus MySQLFilter::onData(Buffer::Instance& data, bool) {
-  doDecode(data);
+  // Safety measure just to make sure that if we have a decoding error we keep going and lose stats.
+  // This can be removed once we are more confident of this code.
+  if (sniffing_) {
+    read_buffer_.add(data);
+    doDecode(read_buffer_);
+  }
   return Network::FilterStatus::Continue;
 }
 
 Network::FilterStatus MySQLFilter::onWrite(Buffer::Instance& data, bool) {
-  doDecode(data);
-  return Network::FilterStatus::Continue;
-}
-
-void MySQLFilter::doDecode(Buffer::Instance& buffer) {
   // Safety measure just to make sure that if we have a decoding error we keep going and lose stats.
   // This can be removed once we are more confident of this code.
-  if (!sniffing_) {
-    buffer.drain(buffer.length());
-    return;
+  if (sniffing_) {
```

Comment:
could you remind me how this sniffing works? the github diff is not showing the entire thing.. I remember this was a boolean that indicated whether we had a decoding error or not and if there was an error, it would simply forward all data as is. But you seem to have flipped the condition.. why?

Reply:
If there was a decoding error, we used to empty the `buffer`. That isn't correct. We need to retain the information in the buffer as that's what is being passed to the next filter (and eventually to MySQL).

In this new logic, I'm not modifying the original `buffer` instance at all. I'm only copying its contents to a new `read_buffer_`/`write_buffer_` instance if the `sniffing_` flag is set to `true`. I've flipped the logic to just reorganize the code.

Reply:
Now, if there's a decoding error, nothing will be copied over to the read/write buffers; however, the original contents of the message will still be forwarded to the next filter hoping that MySQL will be able to handle it correctly.